### DISCUSSION TITLE
How to interface a rc522 rfid module with the SPI of PolarFire SoC Icicle Kit

### DISCUSSION THREAD

**Discussion by premjithav** [polarfire-soc/polarfire-soc-documentation#293] [Linux drivers for PolarFire Soc]
For SPI testing, We tried to enable the spidev driver in Kernel and also enabled in DTS, but the driver is not probing at all. Please suggest a method to interface rc522 with PolarFire SoC Icicle Kit.

**Comment by hughbreslin** [comment]
Hey @premjithav what interface are you using for SPI? The Mikrobus or RPI? In the reference design the only SPI enabled is SPI1 connected to the Mikrobus so just to sanity check you are using that SPI.

**Reply by premjithav** [reply]
Hi, Thanks for the reply. We are using the SPI interface in mikrobus, The spidev test driver in Linux is now enabled and when we tried the loopback test (MISO, MOSI shorted) using spidev_test utility, getting the zero results like below. Please help us to solve this issue

> ./spidev_test -v
> spi mode: 0x0
> bits per word: 8
> max speed: 500000 Hz (500 KHz)
> TX | FF FF FF FF FF FF 40 00 00 00 00 95 FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF F0 0D  | ......@....?..................?.
> RX | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  | ................................

**Reply by hughbreslin** [reply]
I have a few things it would be useful to check / test:
1. If you are using our base XML, you could try changing the SPI I/O configuration to use pull down resistors with a higher drive strength:
![image](https://user-images.githubusercontent.com/44862881/221136297-6af5fbb8-baf1-4f0a-8b47-ac4005991852.png)
You will need to re-build the HSS with updated XML to implement this change
2. Are you just using the standard chip select out of MikroBus?
3. Looking at your output is SPI configured to run at 0.5MHz? I'm asking as there is a limit on the max speed of the SPI controller, I think 20MHz for master, if you exceed this you see data corruption

**Reply by premjithav** [reply]
We tried to Pull down the SPI pins and the max speed settings as suggested by the above reference image, but still the spi loopback test output is same. Is there any  baremetal/linux program available from your side to test the SPI pins in Mikrobus . Also check the attached ZIP file contains the ICICLE xml/cfg files for reference.
[ICICLE_MSS.zip](https://github.com/polarfire-soc/polarfire-soc-documentation/files/11484035/ICICLE_MSS.zip)

**Reply by nitindeshpande** [reply]
@premjithav 
Is this [loopback example](https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/tree/main/driver-examples/mss/mss-spi/mpfs-spi-master-slave) of help?
This project can be used with the SPI_LOOPBACK variant of the [PolarFire SoC reference design](https://mi-v-ecosystem.github.io/redirects/repo-icicle-kit-reference-design)
